tachyon/crypto/sumcheck/sumcheck_verifier.h

```diff
@@ -0,0 +1,204 @@
+  std::vector<F> point;
+  // The expected evaluation
+  F expected_evaluation;
+};
+
+template <typename MLE>
+class SumcheckVerifier {
+ public:
+  using F = typename MLE::Field;
+
+  SumcheckVerifier() = default;
+  explicit SumcheckVerifier(const SumcheckVerifyingKey& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+
+  // Runs a "round" given a |SumcheckProverMsg|, sampling and storing a random
+  // value.
+  //
+  // While a true verifier round should perform actual verification, |Round()|
+  // merely samples and stores random values. Verifications will be performed
+  // altogether in |CheckAndGenerateSubclaim| after all prover and verifier
+  // sub-rounds are subsequently finished.
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // |random_value| is the challenge that will given from the Verifier
+  // to the Prover aka rᵢ
+  SumcheckVerifierMsg<F> Round(SumcheckProverMsg<F>&& prover_msg) {
+    CHECK(!finished_);
+
+    F random_value = F::Random();
+    randomness_.push_back(random_value);
+    polynomials_received_.push_back(prover_msg.evaluations);
+
+    if (round_ == num_variables_) {
+      finished_ = true;
+    } else {
+      ++round_;
+    }
+    return {std::move(random_value)};
+  }
+
+  // Verifies the sumcheck phases' validity and generates a subclaim
+  //
+  // If the asserted sum is correct, then the multilinear polynomial evaluated
+  // at |subclaim.point| is |subclaim.expected_evaluation|. Otherwise, it is
+  // highly unlikely that these two will be equal. A larger field size
+  // guarantees a smaller soundness error.
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // |CheckAndGenerateSubclaim()| verifies for each round i:
+  //   gᵢ(0) + gᵢ(1) =  gᵢ₋₁(rᵢ₋₁)
+  // Note: For the first round, gᵢ₋₁(rᵢ₋₁) = H = |LinearCombination.Combine()|
+  //   Meanwhile, for the last round, gᵢ₋₁(rᵢ₋₁) = the |LinearCombination|
+  //   evaluated on the multivariate point of Verifier's challenges.
+  bool CheckAndGenerateSubclaim(F& asserted_sum, Subclaim<F>& result) {
+    // Verifier has not finished
+    CHECK(finished_);
+    // Insufficient rounds
+    CHECK_EQ(polynomials_received_.size(), num_variables_);
+
+    F expected = asserted_sum;
+    for (size_t i = 0; i < num_variables_; ++i) {
+      std::vector<F>& evaluations = polynomials_received_[i];
+      // Incorrect number of evaluations
+      CHECK_EQ(evaluations.size(), max_evaluations_ + 1);
+
+      F p0 = evaluations[0];
+      F p1 = evaluations[1];
+
+      F sum = p0 + p1;
+      // Check that prover message is consistent with the claim
+      if (sum != expected) {
+        return false;
+      }
+
+      expected = InterpolateUniPoly(evaluations, randomness_[i]);
+    }
+    result = {randomness_, std::move(expected)};
```

Comment:
Please put const or modify this to `result = {std::moev(randomness_), std::move(expected)}`